numpy/core/src/umath/ufunc_type_resolution.c

```diff
@@ -1755,6 +1755,21 @@ ufunc_loop_matches(PyUFuncObject *self,
             if (tmp == NULL) {
                 return -1;
             }
+            if (strcmp(self->name, "floor_divide") == 0 &&
+                PyArray_DESCR(op[0])->type_num == 7 &&
+                PyArray_DESCR(op[1])->type_num == 8) {
```

Comment:
I did a pretty detailed audit of the control flow. The shortest summary I can give is that  `ufunc->ntypes` jumps to `21` in the new behavior vs. `20` in the original case. That was the first point of divergence I found. This seems to empower identification / registration of a floordiv inner loop we don't want to allow in specific cases like `int64 //= uint64`, but do want to allow for `int64 // uint64`.

Our docs describe `ntypes` as:
> ... how many different 1-d loops (of the builtin data types) are available.

As far as I can tell the machinery is already checking type codes or characters in `ufunc_loop_matches`. This function checks both operand types and the output type according to the "cancast" machinery. We've registered a new int64 output type (mm->q) for floordiv, but we want to prohibit it in some cases.

Reply:
> but we want to prohibit it in some cases.

I'm not sure if this is true - see my comment on the issue itself. Allowing `int64 //= uint64` is a logical consequence of our casting rules, and is primarily caused by allowing `uint64 -> timedelta64` casting (unsigned to signed) but not `uint64 -> int64` casting

Reply:
I'm happy to make whatever change is desired. We definitely need tests to guard these casting rules though, otherwise they can and will break as we've seen here.